wordpress-mobile/WordPress-iOS
# Media: Implement picker mode

Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/21457

- Rename `MediaViewController` to `SiteMediaViewController` (more clear what it's for)
- Extract `SiteMediaCollectionViewController` to manage the media collection and use it in a new `SiteMediaViewController`
- Add `SiteMediaPickerViewController`
- `SiteMediaViewController` now displays the number of selected items in the toolbar

I've been writing ViewModels with reactive programming for the past five years, so I'm going old-school here with full MVC 😄 I think it's easier to understand when `NSFetchResultsController` is involved.

<img width="320" alt="media" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/8e68ca80-0334-4442-8467-6a3e06b7101e"> <img width="320" alt="picker" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/849457ab-288c-46c9-8579-a9182aa81596">

## To test:

We don't have any particular requirements for this screen, and it's a bit early to test until we get to phase 2. But if you want to see it, add this somewhere:

// Single selection

```swift
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(2)) {
            let vc = SiteMediaPickerViewController(blog: self.blog!)
            let nav = UINavigationController(rootViewController: vc)
            self.present(nav, animated: true)
        }
```

// Video with multiple selection

```swift
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(2)) {
            let vc = SiteMediaPickerViewController(blog: self.blog!, filter: [.video], allowsMultipleSelection: true)
            let nav = UINavigationController(rootViewController: vc)
            self.present(nav, animated: true)
        }
```

- Verify that the filter works
- Verify that search works
- Verify that selection works and updates the state of the "Done" button
- Verify that single selection works (nothing will happen; it calls the delegate)
- Verify that "Cancel" works (nothing will happen; it calls the delegate)
- Verify that it doesn't show pending uploads

## Regression Notes
1. Potential unintended areas of impact: none
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual
3. What automated tests I added (or what prevented me from doing so): none

PR submission checklist:


UI Changes testing checklist:
